File: 13. Rekursi Lanjut/C. Permutasi Zig-Zag/c_permutasi_zig_zag.py
def write(depth):
    if depth > N:
        isZigZag = True
        if N >= 3:
            for i in range(2, N):
                h = save[i] > save[i-1] and save[i] > save[i + 1]
                l = save[i] < save[i-1] and save[i] < save[i + 1]
                if not (h or l):
                    isZigZag = False
                    break
        if N < 3 or isZigZag:
            for i in range(1, N+1):
                print(save[i], end='')
            print()
    else:
        for i in range(1, N+1):
            if not isVisited[i]:
                isVisited[i] = True
                save[depth] = i
                write(depth+1)
                isVisited[i] = False


isVisited = [False] * 10
save = [0] * 10
N = int(input())
write(1)

File: 13. Rekursi Lanjut/C. Permutasi Zig-Zag/c_permutasi_zig_zag.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

int N;
bool isVisited[17];
int save[17];

void write(int depth){
    if (depth > N){
        bool isZigZag = true;
        if (N >= 3){
            for (int i = 2; i < N; i++){
                bool h = (save[i] > save[i-1]) && (save[i] > save[i + 1]);
                bool l = (save[i] < save[i - 1]) && (save[i] < save[i + 1]);
                if (!(h || l)){
                    isZigZag = false;
                    break;
                }
            }
        }
        if (N < 3 || isZigZag){
            for (int i = 1; i <= N; i++){
                printf("%d", save[i]);
            }
            printf("\n");
        }
    }else{
        for (int i = 1; i <= N; i++){
            if (!isVisited[i]){
                isVisited[i] = true;
                save[depth] = i;
                write(depth + 1);
                isVisited[i] = false;
            }
        }
    }
}

int main(){
    scanf("%d", &N);
    memset(isVisited, false, sizeof(isVisited));

    write(1);
    return 0;
}
